crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1355,6 +1392,111 @@ mod tests {
             ]
         );
     }
+
+    #[test]
+    fn test_jupyter_magic_help_end() {
+        let source = r"
+?foo?
+??   foo?
+??   foo  ?
+?foo??
+??foo??
+???foo?
+???foo??
+??foo???
+???foo???
+?? \
+    foo?
+?? \
+?
+????
+%foo?
+%foo??
+%%foo???"
```

Comment:
What about `%?foo??` or `%???`

Reply:
Oh this is interesting and weird. Let me recap and get the semantics verified from the official implementation:

* Only 1 or 2 question marks at the end are valid and considered otherwise whichever token is present at the start is used to determine the `MagicKind`
* If there are 1 or 2 question marks at the end, only the `%`/`%%` token is _included_ in the command value, other magic kind tokens are ignored. But the magic kind will still be `Help`/`Help2` as it's still a priority:
	* `%foo?` -> (`%foo`, `Help`)
	* `!foo?` -> (`foo`, `Help`)
	* `/foo?` -> (`foo`, `Help`)
* Now, here's where things get weird. Looking at your example and remembering that the implementation is a regex:
	* `%?foo??` is a help command with value `foo`. Here, as `%` is followed by `?` it's ignored. On this note, any question mark in the middle will make every character on and before that position being ignored i.e., `%foo?bar?` will be a help command with value `bar` (again, remember it's a regex with captures).
	* `%???` is a magic command (`%`) as there are more than 2 question marks at the end

Reply:
My main question / concern is what is the correct implementation and should we follow the IPython implementation to the mark? If so, we'll have to make a few small changes to account for the above points.

Reply:
My examples are probably purely artificial. We can decide to ignore them until they get reported.